### ISSUE TITLE
Reduce Memory Requirements

### ISSUE THREAD

**Issue by ranocha** [JuliaDiffEq/OrdinaryDiffEq.jl#178]
This came up in #177.

> The memory consumption of `SSPRK22`, `SSPRK33`, `SSPRK432`, and `SSPRK932` could be reduced by removing `tmp::uType` in the corresponding caches. However, this temporary storage is used in https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/22b52572ebc3d03d2d42685c9cdd340a0d704511/src/callbacks.jl#L62 
Thus, it might be useful to use another temporary storage for the callbacks, since memory consumption can be problematic for big problems / on the GPU.

@ChrisRackauckas:
> Yes, this is a larger issue. Quite RAM I've just been using storage in order to make sure everything is pre-allocated, and not making sure it's minimal. I have opened up issues for lots of places that can be improved (Runge-Kutta methods can get a big improvement through aliasing), but it hasn't been something I've gotten back to since, at least for what I do, RAM is so cheap. But yes, we should clearly document where it needs to be decreased, especially for "optimal storage" type methods, and weed this out. Another place that should be "optional" is saving the prototypes: rate_prototype etc. Those are probably unnecessary after they are used to generate the cache, but I think might get saved to the integrator.

Some issues that have to be considered:
- [x] Low-storage methods might be used to compute the solution only at the final time. Here, no interpolations (dense output, saving `uprev` and derivative etc.) are necessary.
- [x] In some (Runge-Kutta) methods, some temporary arrays could be reused.
  > Runge-Kutta methods can get a big improvement through aliasing
- [x] Saving prototypes only if necessary.
  > Another place that should be "optional" is saving the prototypes: rate_prototype etc. Those are probably unnecessary after they are used to generate the cache, but I think might get saved to the integrator.
- [ ] Add low storage schemes with specialised implementations:
  - http://epubs.siam.org/doi/abs/10.1137/07070485X. Some SSP methods such as `SSPRK104` are already implemented (but with potentially higher memory consumption, see above).
  - http://www.sciencedirect.com/science/article/pii/S0021999109006251
  - https://www.sciencedirect.com/science/article/pii/S0021999104001913
  - http://www.sciencedirect.com/science/article/pii/S0021999111005213
  - http://www.sciencedirect.com/science/article/pii/S0168927499001415
  - http://www.sciencedirect.com/science/article/pii/0021999180900339
  - https://ntrs.nasa.gov/search.jsp?R=19940028444
- [x] Supporting units can lead to more memory usage.

However, reducing storage requirements might conflict with the generality of all use cases (dense output, callbacks, etc).

**Comment by ChrisRackauckas** [comment]
Crossref: https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/issues/17

**Comment by ChrisRackauckas** [comment]
Add to the list that supporting units can lead to more cache usage.

I think this should all be handled in the cache and we can be a bit more liberal with aliasing there. If the `alg_cache` function has the right information, we can do things like "check if `one(u) == oneunit(u)` and if so, make `tmp2` different from `tmp`, otherwise make them alias". That way the `perform_step` and everything else is the same, but we can detect cases where we can use less memory. Or even for some algorithms like the RK algorithms, it's fine to not have say `k3` at the end of the algorithm (it's only used in the first 5 steps, and not in the last 2 or in the dense output). I think relying on aliasing like this can be easier than the Fortran style of never aliasing, which leads to this kind of code in dopri:

```fortran
      DO 28 I=1,N 
  28  K4(I)=(E1*K1(I)+E3*K3(I)+E4*K4(I)+E5*K5(I)+E6*K6(I)+E7*K2(I))*H
```

where it's doing what I said above and effectively using `K7 === K2`, but then it gets hard to follow if you don't recognize that. The downside to this approach is that if you don't know the cache the `perform_step` can do weird things if you aren't aware two arrays alias. But I think tests would catch any error like that, and I think it's more legible in the long run?

To save some memory, we should just take out 

https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/src/integrators/type.jl#L54

and see if tests pass. My guess is it will work. I forget why I added it in and I have been meaning to get rid of it and make it part of the cache.

Other than that, in the integrator we have

https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/src/integrators/type.jl#L51

which only gets allocated for `alg_extrapolates`, so that's fine. But

https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/src/integrators/type.jl#L50

we can be smarter with `u` vs `uprev`. If `adaptive==false`, `calck=false` (essentially `dense=false`), and there are no callbacks, we can alias these (and then the fallback dense output would end up as piecewise constant instead of linear though, which should be fine if documented appropriately). 

All other allocations are in caches and can be handled algorithm by algorithm. 

@devmotion do you think this sounds like a good idea?

**Comment by devmotion** [comment]
> Add to the list that supporting units can lead to more cache usage.

I already noticed that, so e.g. for calculating residuals additional unitless arrays are required - but as you said, we could just alias those whenever e.g. `u` is unitless. But it needs some care.

> To save some memory, we should just take out
https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/src/integrators/type.jl#L54
and see if tests pass. My guess is it will work. I forget why I added it in and I have been meaning to get rid of it and make it part of the cache.

I also guess that it will work, can't remember any occurrences or requirements past the initialization phase.

> The downside to this approach is that if you don't know the cache the perform_step can do weird things if you aren't aware two arrays alias. But I think tests would catch any error like that, and I think it's more legible in the long run?

I had some problems in DelayDiffEq because of unnoticed array aliases, but nevertheless I think it's good to get rid of unnecessary arrays. Although most errors should be catched by tests I guess we should document any changes, especially if they're not trivial and/or deviate from literature or usual nomenclature.

**Comment by YingboMa** [comment]
Note that almost all the Runge-Kutta-Nyström and symplectic solvers have superfluous caches too.

**Comment by ChrisRackauckas** [comment]
@YingboMa could you expand on that?

It looks like we have the tools necessary to handle this. I went ahead and demo'd it on `DP5()`.

https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/src/caches/low_order_rk_caches.jl#L437-L463

and added a specific test to OrdinaryDiffEqExtendedTests.jl to make sure it doesn't alter the calculation in this case.

https://github.com/JuliaDiffEq/OrdinaryDiffEqExtendedTests.jl/commit/427a29702383b2d093ee14744487c07f6f96417e

The gain is 5 aliases in the standard case and 7 when no units or interpolation are used. That brings it to 7 and 9 registers respectively. This shows it's not a structural limitation or anything of OrdinaryDiffEq.jl, we just need to go do this to the other methods as well.

So the user can always get a "fast setting" by turning off `calck` when they don't need it. The issue is how to default `calck`. We can do something quite smart here to catch just about every use case. We know that interpolations are only used when one of these is true:

- `saveat` is used at points not in `tstops`
- `dense` is used
- there's a continuous callback
- the user manually uses the integrator interpolation in a discrete callback
- the user manually uses the integrator interpolation in the integrator interface

I think we should go with a relatively "safe" default since it's easy to turn off but hard to debug if incorrect (since all you'd see is that the interpolations would calculate something, but it would be wrong. We can make this error instead?). The safe default would be:

- Add a keyword argument if `solve` is used instead of `init`. This would allow us to know when the user is manually using the integrator interface
- Default calck to true unless `isempty(setdiff(saveat,tstops))`, `!dense`, there are no callbacks, and we detect that `init` is not directly used instead of through `solve`.

This makes the standard case of solving a PDE with `dense=false` (or `save_everystep=false`) catch the fast route perfectly fine. Another thing we can do is make this a little algorithm-dependent, as in for low-storage methods just assume `calck=false` by default (but still have a route with an interpolation?) even when callbacks or integrator interface is used, since it's very unlikely you're using a low storage method for some interpolation purpose. I'd want to have an error with integrator interpolations if `calck=false` if we're going to do that though.

But yeah, the key is here's an example of it being done, so now we just need better defaults, more visibility for the `calck` argument, and more implementations which make use of this.

**Comment by ChrisRackauckas** [comment]
I checked

> Low-storage methods might be used to compute the solution only at the final time. Here, no interpolations (dense output, saving uprev and derivative etc.) are necessary.

since that's redundant with optimizing the memory of the RK methods. The memory optimized DP5 shows how you can disrupt the interpolation to minimize memory and this is fine. We have the right flags (`calck=false` which is set by things like `save_everystep=false`) so that the user can tell us this is okay.

In fact, extra memory due to units is similar in that we can alias arrays in the non-units case (seen in DP5 as well).

So really the concrete steps are to do this full aliasing like in DP5 but in every cache, and implement low-storage methods.

**Comment by ChrisRackauckas** [comment]
@ranocha do you have a good test to find extra memory requires? With the prototype storage removed and the units handled by aliasing when non-unitful, making things require less memory isn't package wide anymore and would be on an algorithm-by-algorithm basis, with many of the main methods already being optimized. We should of course find a way to benchmark this, but nothing about the structure of DiffEq itself has extra memory requirements anymore, just some algorithm cache implementations.

**Comment by ranocha** [comment]
No, I don't have any good general test to find extra memory requires. There are some specialised routines that can be used, but most of them would need additional function evaluation routines provided by the user. Since this is a more general suite of ODE solvers, it might not be worth the effort.

**Comment by ChrisRackauckas** [comment]
Well, this isn't a general suite in the traditional sense, it's a general suite of specific methods. For the methods which are low memory, we should make sure that there are 0 extra arrays copies of the state array.

**Comment by ranocha** [comment]
Most low-memory methods rely on some assumptions such as "We can evaluate `u2 = u1 + dt * f(u1)` using only memory for two variables `u1` and `u2`. We would have to mess up with or alias) `k`s and `u`s.

**Comment by ranocha** [comment]
I have to think about possible tests. One problem is that the minimal memory requirement depends on the method at hand. And I don't really know what might be the best way to test the number of array allocations.
